docs/architecture/adr-017-subspace-fee-grant.md

```diff
@@ -0,0 +1,261 @@
+
+The new subspace-specified `DeductFeeDecorator` will work as follows:
+1. check if all the messages in the transaction are messages related to the same subspace;
+2. if the messages in the transaction are all realted to the same subspace and a fee grant exists for that subspace, run the `x/subspaces` `DeductFeeDecorator`; otherwise run the `x/auth` `DeductFeeDecorator`,
+
+### Types
+
+#### Allowance
+
+The `x/feegrant` module already provides an interface (`FeeAllowanceI`) to represent a generic allowance, along with a set of useful implementations like `BasicAllowance`, `AllowedMsgAllowance` and `PeriodicAllowance`. Since we are not going to introduce new kinds of allowances, nor we are going to edit how an allowance is represented, we will reuse these types.
+
+#### User Grant
+
+The `x/feegrant` module provides a `Grant` object used to store the `granter`, `grantee` and what kind of `allowance` is granted to a specific user. Since this is enough information for us as well, we will use the same object to store user-related fee grants.
+
+#### Group Grant
+
+In order to represent a fee grant granted to a user group, we will implement a `GroupGrant` type. This will contain the `granter`, `group_id` and what kind of `allowance` is granted to the group:
+
+```protobuf
+message GroupGrant {
+    // granter is the address of the user granting an allowance of their funds.
+  string granter = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+
+  // the id of the group being granted an allowance of another user's funds.
+  uint32 group_id = 2;
+
+  // allowance can be any of basic, periodic, allowed fee allowance.
+  google.protobuf.Any allowance = 3 [(cosmos_proto.accepts_interface) = "FeeAllowanceI"];
+}
+```
+
+### Store
+
+#### Group grant
+
+In order to simplify granters to manage the group allowance, the granted group allowance will be stored in the keys having the structure like:
+```
+SubspaceGroupAllowancePrefix | SubspaceID | GranterAddress | GroupID | -> Protobuf(GroupGrant)
+```
+
+This structure allows granters to easily manage the group allowance inside a subspace by iterating over all allowances for the granters, which will be the most used query. In the other hand, grantees must know who the granter is when using the application, they can find their grant with O(N) time complexity, N is the number of granted groups by the granter.
+
+#### User grant
+
+Each subspace fee granted allowance will be stored in the keys having the structure as follows:
+```
+SubspaceUserAllowancePrefix | SubspaceID | GranterAddress | GranteeAddress |-> Protobuf(Grant)
+```
+
+This structure allows granters to easily manage their grants inside a subspace by iterating over all grants for the granters, which will be the most used query. In the other hand, grantees must know who the granter is when using the application, they can directly find their grant with O(1) time complexity.
```

Comment:
```suggestion
This structure allows granters to easily manage their grants inside a subspace by iterating over all grants for the granters, which will be the most used query. On the other hand, it allows the following queries:

1. retrieving a specific grant.  
   This will require grantees to know who the granter address, and can be done in O(1).
2. retrieving all the grants that a user has been granted. 
   This can be done by iterating over all the granters' grants in O(N * M), where N is the number of granters and M is the number of grants each granter has granted.
```

Reply:
@dadamu But what if the user does not know the granter address, or wants the overall grants that have been granted? In that case we need to iterate over the prefix `SubspaceUserAllowancePrefix | SubspaceID` and check if each grant is referring to the user or not

Reply:
If the user does not know the granter, it would be O(N), where N is the number of all the grants.
However, from my side, the grantee must know who is the granter. Assume that there is an application provided by a granter, the granter would always grant new users subspace allowance, then users use that application to operate inside the subspace must know who the application granter.

Reply:
@RiccardoM If the user does not know the granter, it would be O(N), where N is the number of all the grants.
However, from my side, the grantee must know who is the granter. Assume that there is an application provided by a granter, the granter would always grant new users subspace allowance, then users use that application to operate inside the subspace must know who the application granter.

Reply:
@dadamu Ok, yeah O(N) with N number of grants looks decent anyway. 

> then users use that application to operate inside the subspace must know who the application granter

Not necessarily, there might be multiple admins that are allowed to create a user-group fee grant. In that case, the user would not not the granter. Or they might have multiple grants because they are inside multiple groups. I think we should not make it necessary for users to know their granter. Also becuase this is a query that does not consume gas and can be implemented quite easily. I don't see any harm in allowing it